packages/teleterm/src/ui/DocumentGateway/useDocumentGateway.ts

```diff
@@ -32,12 +33,14 @@ export default function useGateway(doc: types.DocumentGateway) {
   const cluster = ctx.clustersService.findClusterByResource(doc.targetUri);
 
   const [connectAttempt, createGateway] = useAsync(async () => {
-    const gw = await ctx.clustersService.createGateway({
-      targetUri: doc.targetUri,
-      port: doc.port,
-      user: doc.targetUser,
-      subresource_name: doc.targetSubresourceName,
-    });
+    const gw = await retryWithRelogin(ctx, doc.uri, doc.targetUri, () =>
```

Comment:
We might be dealing with a race condition here, but now that I wrote it down I think we're fine.

Once you click "Connect" next to a db and pick the db user and db name, this code is executed:

https://github.com/gravitational/webapps/blob/627dea4ad701be3c5fae36e06ee73da4837d855c/packages/teleterm/src/ui/DocumentCluster/ClusterResources/Databases/useDatabases.ts#L41-L42

This adds the document, which in turn fires this `useAsync` hook and we fire a request to the tsh daemon. If we get an error related to expired certs, there's a check in `retryWithRelogin` to see if the user is still seeing the relevant UI element.

So we could think of a situation where `documentsService.add` is fired and before `documentsService.open` is executed, we already got a cert expired error from the gateway, but since we haven't open the doc yet, the check in `retryWithRelogin` will think that the user is not looking at the relevant doc.

However, as far as I understand JavaScript's runtime, both `add` and `open` are not async functions. This means that any async code executed inside them will happen on the next tick, after the whole `useDatabases.connect` function finishes its execution.

So I think we should be fine here, even in theory? Because in practice it's obviously hard for the daemon to respond before we open the doc.

Reply:
Correct. The JS engine will not execute the asynchronous tasks before the synchronous ones finish.